### PR TITLE
ci(build): Improve pull request build

### PR DESCRIPTION
**Pull request by validcube**
# 👷 Improve build workflow

> [!IMPORTANT]
> **Last updated for 1e5833dc12777cee3b202bb942b2bf367c9d0b3a otherwise notice**

`pr-build.yml` will only be triggered when you manually dispatch it, it will comment on the PR about the build status.

the CI will require you to input your PR number, and you can customise the app flavour and flutter variant

By the way, since we are only manually dispatching it, we can finally use Gradle cache, there's less chance to hit the cache eviction compared to our old build CI.

## Overview
The way this work is once dispatched, the CI will build the application with the choice that you've provided, when the process of building the application is done, it will notify the owner of the Pull Request to tell them if there's a problem with their contributions or not.

### Artefact
If the application is built without any issues. The CI will upload the APK and SHA hash as artefact under `revanced-manager` App name will always come before the bracket, the bracket is a group that indicates everything inside was build-related, and the outside was tool-related.

**revanced-manager-(e8f044d-17-release)-stable**

image

### Notify
image

The CI will comment on the PR notifying them of: the time when the workflow was dispatched, configuration and build status, 
> ## ⚒️ ReVanced PR Build workflow
>
> ✅ The CI has sucessfully built ReVanced Manager at 85608fc.
>
> You can check the details on run ID [6667156871](example.com)!
>
> ### ⚙️ Overview
> - App flavor: release
> - Flutter branch: stable
> - Flutter cache: true
> - Start time: Fri, 27 Oct 2023 12:25:53 +0000

the CI will also try to recreate the comment on every dispatch.

> [!WARNING]
> Don't run the CI more than two instances of the workflow on the same PR or else it might create additional comment instead of recreating it.

> [!NOTE]
> It's possible to use custom account such as bot account to comment by replacing the GitHub Actions token with your own.

pr-build.yml : `notify-pull-request` Job
```yml
...
      - name: "Comment to Pull Request #${{ inputs.pr-number }}"
        uses: thollander/actions-comment-pull-request@v2
        with:
          GITHUB_TOKEN: ${{ secret.bot-token }}
...
```

## Inputs

image

### Supported Flutter branch // `choice`
* Stable
* Beta
* Dev
* Master (why)

* true
* false

### Supported app flavour // `choice`
* Release
* Debug
* [Profile ⧉](https://docs.flutter.dev/perf/ui-performance#run-in-profile-mode:~:text=Flutter%E2%80%99s%20profile%20mode%20compiles%20and%20launches%20your%20application%20almost%20identically%20to%20release%20mode%2C%20but%20with%20just%20enough%20additional%20functionality%20to%20allow%20debugging%20performance%20problems.%20For%20example%2C%20profile%20mode%20provides%20tracing%20information%20to%20the%20profiling%20tools. "Flutter Profile mode")

Pull Request number without the hashtag (i.e 1337, not #1337)

### CODE REVIEW

File: .github/workflows/pr-build.yml [hunk 1]
```diff
@@ -37,9 +64,54 @@ jobs:
...
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - name: Set env
+        run: |
+          echo "COMMIT_HASH=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
+          if [[ "${{ needs.build.outputs.build_status }}" == "success" ]]; then
+            echo "MESSAGE=✅ Build Success!" >> $GITHUB_OUTPUT
+          else
+            echo "MESSAGE=🚫 Build Fail!" >> $GITHUB_OUTPUT
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:89]:
```suggestion
            echo "MESSAGE=✅ Build succeeded!" >> $GITHUB_OUTPUT
          else
            echo "MESSAGE=🚫 Build failed!" >> $GITHUB_OUTPUT
```

File: .github/workflows/pr-build.yml [hunk 2]
```diff
@@ -37,9 +64,54 @@ jobs:
...
+          fi
+      - name: Comment on Pull Request ${{ inputs.pr-number }}
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:91]:
```suggestion
      - name: Comment to pull request ${{ inputs.pr-number }}
```

File: .github/workflows/pr-build.yml [hunk 3]
```diff
@@ -37,9 +64,54 @@ jobs:
...
+              - App variant: ${app_variant}
+              - Flutter variant: ${flutter_variant}
+            `;
+
+            github.issues.createComment({
+              issue_number: pr_number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: comment
+            });
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:117]: You may want to use the PR comment action (See revanced-website)

File: .github/workflows/pr-build.yml [hunk 4]
```diff
@@ -37,9 +64,54 @@ jobs:
...
+        run: |
+          flutter build apk --${{ inputs.app-variant }};
+      - name: Set env
+        run: |
+          echo "COMMIT_HASH=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
       - name: Upload build
         uses: actions/upload-artifact@v3
         with:
-          name: revanced-manager
+          name: revanced-manager-${{ env.COMMIT_HASH }}-P${{ inputs.pr-number }}-B${{ inputs.app-variant }}-F${{ inputs.flutter-variant }}
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:75]: What is the purpose of P, B and F?

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: .github/workflows/pr-build.yml (hunk 4 above)
Comment (validcube) [reply .github/workflows/pr-build.yml:75]: Just a placeholder, will delete soon

- P: Pull Request
- B: Build variant
- F: Flutter variant

File: .github/workflows/pr-build.yml [hunk 5]
```diff
@@ -37,9 +63,39 @@ jobs:
...
+          flutter build apk --${{ inputs.app-variant }};
+      - name: Set env
+        run: |
+          echo "COMMIT_HASH=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
       - name: Upload build
         uses: actions/upload-artifact@v3
         with:
-          name: revanced-manager
+          name: revanced-manager-${{ env.COMMIT_HASH }}-${{ inputs.pr-number }}-${{ inputs.app-variant }}-${{ inputs.flutter-variant }}
           path: build/app/outputs/flutter-apk/app-debug.apk
```
Comment (validcube) [inline .github/workflows/pr-build.yml:75]: TODO: `app-${{ inputs.app-variant }}`

> This will cause upload artefact to fail when compile with **release** and **profile** mode.

File: .github/workflows/pr-build.yml [hunk 6]
```diff
@@ -37,9 +63,39 @@ jobs:
...
+        with:
+          GITHUB_TOKEN: ${{ secrets.PULL_REQUESTS_WRITE }}
+          message: |
+            ${{ env.MESSAGE }} @ ${{ env.COMMIT_HASH }}
+            
+            You can check the details [here](${{ github.event.workflow_run.html_url }}).
+  
+            - PR: ${{ inputs.pr-number }}
+            - App variant: ${{ inputs.app-variant }}
+            - Flutter variant: ${{ inputs.flutter-variant }}
```
Comment (validcube) [inline .github/workflows/pr-build.yml:101]: TODO: This will fail... I forgor to specify PR

File: .github/workflows/pr-build.yml [hunk 7]
```diff
@@ -37,9 +63,39 @@ jobs:
...
           path: build/app/outputs/flutter-apk/app-debug.apk
+          
+  notify-pull-request:
+    name: Notify Pull Request
+    runs-on: ubuntu-latest
+    needs: build
+    steps:
+      - name: Set env
+        run: |
+          echo "COMMIT_HASH=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
```
Comment (validcube) [inline .github/workflows/pr-build.yml:84]: ????

File: .github/workflows/pr-build.yml [hunk 8]
```diff
@@ -37,9 +63,39 @@ jobs:
...
+            echo "MESSAGE=✅ Build succeeded!" >> $GITHUB_OUTPUT
+          else
+            echo "MESSAGE=🚫 Build failed!" >> $GITHUB_OUTPUT
+          fi
+      - name: Comment to Pull Request #${{ inputs.pr-number }}
+        uses: thollander/actions-comment-pull-request@v2
+        with:
+          GITHUB_TOKEN: ${{ secrets.PULL_REQUESTS_WRITE }}
+          message: |
+            ${{ env.MESSAGE }} @ ${{ env.COMMIT_HASH }}
```
Comment (validcube) [inline .github/workflows/pr-build.yml:95]: TODO: Will definitely look super awkward when `COMMIT_HASH` not available

@

**Review by validcube — COMMENTED** [review]
I'll leave a list of things to TODO here

File: .github/workflows/pr-build.yml [hunk 9]
```diff
@@ -1,25 +1,53 @@
...
+    continue-on-error: true
     runs-on: ubuntu-latest
     steps:
+      - name: Checkout the repository
+        env: 
+          GH_TOKEN: ${{ github.token }}
+        run: |
+          gh repo clone validcube/revanced-manager
+          cd revanced-manager
+          gh repo set-default validcube/revanced-manager
```
Comment (validcube) [inline .github/workflows/pr-build.yml:45]:
```suggestion
          gh repo clone ReVanced/revanced-manager
          cd revanced-manager
          gh repo set-default ReVanced/revanced-manager
```

File: .github/workflows/pr-build.yml [hunk 10]
```diff
@@ -1,25 +1,53 @@
...
+        - master
+      app-variant:
+        required: true
+        type: choice
+        description: App variant
+        default: 'release'
+        options: 
+        - release
+        - debug
+        - profile
```
Comment (validcube) [inline .github/workflows/pr-build.yml:24]:
```suggestion
```

Known to be problematic

**Comment by validcube** [comment]
Whoops, didn't mean to mark as ready

File: .github/workflows/pr-build.yml [hunk 11]
```diff
@@ -37,9 +66,41 @@ jobs:
...
+      - name: "Comment to Pull Request #${{ inputs.pr-number }}"
+        uses: thollander/actions-comment-pull-request@v2
+        with:
+          GITHUB_TOKEN: ${{ github.token }}
+          pr_number: ${{ inputs.pr-number }}
+          mode: recreate
+          message: |
+            ${{ env.MESSAGE }} @ ${{ env.DATETIME }}
+            
+            You can check the details [here](${{ github.event.workflow_run.html_url }}).
```
Comment (validcube) [inline .github/workflows/pr-build.yml:102]: TODO: Fix URL... ????? HOW

Why is this linked to the PR's URL, not the workflow url

File: .github/workflows/pr-build.yml [hunk 12]
```diff
@@ -1,34 +1,63 @@
...
+          gh repo clone ReVanced/revanced-manager
+          cd revanced-manager
+          gh repo set-default ReVanced/revanced-manager
+          gh pr checkout ${{ inputs.pr-number }}
       - name: Checkout
         uses: actions/checkout@v4
         with:
-          # Make sure the release step uses its own credentials:
-          # https://github.com/cycjimmy/semantic-release-action#private-packages
           persist-credentials: false
```
[the remark this answers is not in the document]
Comment (oSumAtrIX) [reply .github/workflows/pr-build.yml:50]: It should be possible to remove it, because the credentials it uses have all permissions necessary.

File: android/app/build.gradle [hunk 13]
```diff
@@ -70,6 +70,16 @@ android {
...
+        profile {
+            shrinkResources false
+            minifyEnabled false
+            resValue "string", "app_name", "ReVanced Manager Profile"
+            applicationIdSuffix ".profile"
+            signingConfig signingConfigs.debug
+            ndk {
+                abiFilters 'arm64-v8a', 'armeabi-v7a', 'x86_64'
+            }
+        }
```
[the remark this answers is not in the document]
Comment (Ushie) [reply android/app/build.gradle:82]: Why does a `profile` variant exist? Is it needed at all?

File: .github/workflows/pr-build.yml (hunk 11 above)
Comment (Ushie) [reply .github/workflows/pr-build.yml:102]: Where did you see that it's returning the PR's URL?

Do you mind sharing the repository where you're testing this action?

File: .github/workflows/pr-build.yml [hunk 14]
```diff
@@ -1,34 +1,63 @@
...
-      - "android/**"
-      - "assets/**"
-      - "lib/**"
-      
+  workflow_dispatch:
+    inputs:
+      flutter-variant:
+        required: true
+        type: choice
+        description: Flutter variant
```
Comment (validcube) [inline .github/workflows/pr-build.yml:9]:
```suggestion
        description: Flutter branch
```

File: .github/workflows/pr-build.yml [hunk 15]
```diff
@@ -37,9 +66,41 @@ jobs:
...
+    needs: build
+    steps:
+      - name: Set env
+        run: |
+          echo "DATETIME=$(date +'%Y-%m-%d %H:%M:%S')" >> $GITHUB_ENV
+          if [[ "${{ needs.build.result }}" == "failure" ]]; then
+            echo "MESSAGE=🚫 Build failed!" >> $GITHUB_ENV
+          else
+            echo "MESSAGE=✅ Build succeeded!" >> $GITHUB_ENV
+          fi
```
Comment (validcube) [inline .github/workflows/pr-build.yml:92]: TODO: this is always true (or false, maybe)

File: .github/workflows/pr-build.yml [hunk 16]
```diff
@@ -37,9 +66,41 @@ jobs:
...
+      - name: Set env
+        run: |
+          echo "COMMIT_HASH=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
       - name: Upload build
         uses: actions/upload-artifact@v3
         with:
-          name: revanced-manager
-          path: build/app/outputs/flutter-apk/app-debug.apk
+          name: revanced-manager-${{ env.COMMIT_HASH }}-${{ inputs.pr-number }}-${{ inputs.app-variant }}-${{ inputs.flutter-variant }}
+          path: build/app/outputs/flutter-apk/app-${{ inputs.app-variant }}.apk
```
Comment (validcube) [inline .github/workflows/pr-build.yml:78]: TODO: include SHA hash

File: .github/workflows/pr-build.yml [hunk 17]
```diff
@@ -1,34 +1,63 @@
...
-  pull_request:
-    paths: 
-      - ".github/workflows/pr-build.yml"
-      - "android/**"
-      - "assets/**"
-      - "lib/**"
-      
+  workflow_dispatch:
+    inputs:
+      flutter-variant:
```
Comment (validcube) [inline .github/workflows/pr-build.yml:6]:
```suggestion
      flutter-branch:
```

File: .github/workflows/pr-build.yml [hunk 18]
```diff
@@ -1,34 +1,63 @@
...
+        
+permissions:
+  contents: read
+  pull-requests: write
+
 jobs:
   build:
-    name: Build
+    name: Build the application
+    continue-on-error: true
```
Comment (validcube) [inline .github/workflows/pr-build.yml:36]:
```suggestion
```

File: .github/workflows/pr-build.yml [hunk 19]
```diff
@@ -1,34 +1,63 @@
...
         uses: actions/setup-java@v3
         with:
           java-version: '11'
           distribution: 'zulu'
+          cache: gradle
       - name: Setup Flutter
         uses: subosito/flutter-action@v2
         with:
-          channel: 'stable'
+          channel: ${{ inputs.flutter-variant }}
```
Comment (validcube) [inline .github/workflows/pr-build.yml:60]:
```suggestion
          channel: ${{ inputs.flutter-branch }}
```

File: .github/workflows/pr-build.yml [hunk 20]
```diff
@@ -37,9 +66,41 @@ jobs:
...
+          flutter build apk --${{ inputs.app-variant }};
+      - name: Set env
+        run: |
+          echo "COMMIT_HASH=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
       - name: Upload build
         uses: actions/upload-artifact@v3
         with:
-          name: revanced-manager
-          path: build/app/outputs/flutter-apk/app-debug.apk
+          name: revanced-manager-${{ env.COMMIT_HASH }}-${{ inputs.pr-number }}-${{ inputs.app-variant }}-${{ inputs.flutter-variant }}
```
Comment (validcube) [inline .github/workflows/pr-build.yml:77]:
```suggestion
          name: revanced-manager-${{ env.COMMIT_HASH }}-${{ inputs.pr-number }}-${{ inputs.app-variant }}-${{ inputs.flutter-branch }}
```

File: .github/workflows/pr-build.yml [hunk 21]
```diff
@@ -37,9 +66,41 @@ jobs:
...
+          pr_number: ${{ inputs.pr-number }}
+          mode: recreate
+          message: |
+            ${{ env.MESSAGE }} @ ${{ env.DATETIME }}
+            
+            You can check the details [here](${{ github.event.workflow_run.html_url }}).
+  
+            - PR: ${{ inputs.pr-number }}
+            - App variant: ${{ inputs.app-variant }}
+            - Flutter variant: ${{ inputs.flutter-variant }}
```
Comment (validcube) [inline .github/workflows/pr-build.yml:106]:
```suggestion
            - Flutter branch: ${{ inputs.flutter-branch }}
```

File: .github/workflows/pr-build.yml [hunk 22]
```diff
@@ -37,9 +66,41 @@ jobs:
...
+        with:
+          GITHUB_TOKEN: ${{ github.token }}
+          pr_number: ${{ inputs.pr-number }}
+          mode: recreate
+          message: |
+            ${{ env.MESSAGE }} @ ${{ env.DATETIME }}
+            
+            You can check the details [here](${{ github.event.workflow_run.html_url }}).
+  
+            - PR: ${{ inputs.pr-number }}
```
Comment (validcube) [inline .github/workflows/pr-build.yml:104]:
```suggestion
```

[the remark this answers is not in the document]
Comment (validcube) [reply .github/workflows/pr-build.yml:104]: Fair enough, I just realised it.

File: .github/workflows/pr-build.yml (hunk 11 above)
Comment (validcube) [reply .github/workflows/pr-build.yml:102]: https://github.com/validcube/revanced-manager/pull/17#issuecomment-1739217455

**Review by validcube — COMMENTED** [review]
Mobile editing !!!

File: .github/workflows/pr-build.yml (hunk 16 above)
Comment (Ushie) [reply .github/workflows/pr-build.yml:78]: SHA isn't really necessary

Comment (validcube) [reply .github/workflows/pr-build.yml:78]: Eh, it's easy to do, so I figure why not add it

```suggestion
          path: | 
            build/app/outputs/flutter-apk/app-${{ inputs.app-variant }}.apk
            build/app/outputs/flutter-apk/app-${{ inputs.app-variant }}.apk.sha1
```

File: android/app/build.gradle (hunk 13 above)
[the remark this answers is not in the document]
Comment (validcube) [reply android/app/build.gradle:82]: > Why does a `profile` variant exist? Is it needed at all?

to be honest, not really but kind of useful for some case when analysing performance issues.

File: .github/workflows/pr-build.yml (hunk 15 above)
Comment (Ushie) [reply .github/workflows/pr-build.yml:93]: Has this been resolved?

File: .github/workflows/pr-build.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ushie) [reply .github/workflows/pr-build.yml:49]: Has this been resolved?

[the remark this answers is not in the document]
Comment (validcube) [reply .github/workflows/pr-build.yml:49]: > Has this been resolved?

Not yet, Palm also have no idea why removing this step failed the job.

File: .github/workflows/pr-build.yml (hunk 15 above)
Comment (validcube) [reply .github/workflows/pr-build.yml:93]: > Has this been resolved?

I pushed a (unverified) fix, that stop the CI when error has occurred.

File: .github/workflows/pr-build.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (validcube) [reply .github/workflows/pr-build.yml:49]: It failed at 

image

with code 1

**Comment by Ushie** [comment]
Is this ready for review?

**Comment by validcube** [comment]
> Is this ready for review?

I'm not sure how to fix the problems that was mentioned in the conversation, any alternative way of implementing this would be appreciated.

File: .github/workflows/pr-build.yml (hunk 15 above)
Comment (oSumAtrIX) [reply .github/workflows/pr-build.yml:93]: Can you check if the workflow pushed an asset?

Comment (validcube) [reply .github/workflows/pr-build.yml:94]: > Can you check if the workflow pushed an asset?

Artifact? Yep.

* APK
* Sha1

**Comment by oSumAtrIX** [comment]
I think a work around is to report that CI finished instead of telling if it succeeded. Alternatively, you can use GitHub CLI to get the status of CI maybe.

File: .github/workflows/pr-build.yml [hunk 23]
```diff
@@ -1,34 +1,63 @@
...
+permissions:
+  contents: read
+  pull-requests: write
+
 jobs:
   build:
-    name: Build
+    name: Build the application
     runs-on: ubuntu-latest
+    continue-on-error: true
```
Comment (validcube) [inline .github/workflows/pr-build.yml:37]: AH HA, THE CULPRIT 

```suggestion
    continue-on-error: false
```

File: .github/workflows/pr-build.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (validcube) [reply .github/workflows/pr-build.yml:58]: Decided to leave it in there using the name "Setup"

File: .github/workflows/pr-build.yml (hunk 15 above)
Comment (validcube) [reply .github/workflows/pr-build.yml]: Fixed in latest revision

image

**Comment by validcube** [comment]
> I think a work around is to report that CI finished instead of telling if it succeeded. Alternatively, you can use GitHub CLI to get the status of CI maybe.

Fixed, I consolidate job and get the status via ${{ steps.build.outcome }}

File: .github/workflows/pr-build.yml [hunk 24]
```diff
@@ -1,45 +1,122 @@
...
+            - Flutter branch: ${{ inputs.flutter-branch }}
+            - Flutter cache: ${{ inputs.flutter-cache }}
+            - Start time: ${{ env.DATETIME }}
+
       - name: Upload build
         uses: actions/upload-artifact@v3
         with:
-          name: revanced-manager
-          path: build/app/outputs/flutter-apk/app-debug.apk
+          error: true
```
Comment (validcube) [inline .github/workflows/pr-build.yml:118]: Ensure that the CI report error when build failed, instead of success because we continue on error

Comment (validcube) [reply .github/workflows/pr-build.yml:118]:
```suggestion
          if-no-files-found: error
```

File: .github/workflows/pr-build.yml [hunk 25]
```diff
@@ -1,45 +1,122 @@
...
-        run: flutter build apk --debug
+        run: |
+          flutter build apk --${{ inputs.app-flavour }};
+          
+      - name: Prepare to comment
+        run: |
+          if [[ "${{ steps.flutter-build.outcome }}" == "success" ]]; then
+            echo "MESSAGE=✅ The CI has sucessfully built ReVanced Manager at ${{ env.COMMIT_HASH }}." >> $GITHUB_ENV
+          else
+            echo "MESSAGE=🚫 The CI has unable to build ReVanced Manager at ${{ env.COMMIT_HASH }}." >> $GITHUB_ENV
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:93]:
```suggestion
            echo "MESSAGE=✅ ReVanced Manager ${{ env.COMMIT_HASH }} build succeeded." >> $GITHUB_ENV
          else
            echo "MESSAGE=🚫 ReVanced Manager ${{ env.COMMIT_HASH }} build failed." >> $GITHUB_ENV
```

File: .github/workflows/pr-build.yml [hunk 26]
```diff
@@ -1,45 +1,122 @@
...
+        with:
+          GITHUB_TOKEN: ${{ github.token }}
+          pr_number: ${{ inputs.pr-number }}
+          mode: recreate
+          message: |
+            ## ⚒️ ReVanced PR Build workflow
+            
+            ${{ env.MESSAGE }}
+            
+            You can check the details on run ID [${{ github.run_id }}](https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }})!
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:107]:
```suggestion
            You can see more details in run [${{ github.run_id }}](https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }})!
```

**Review by oSumAtrIX — APPROVED** [review]

File: .github/workflows/pr-build.yml [hunk 27]
```diff
@@ -1,45 +1,122 @@
...
+        default: 'stable'
+        options: 
+          - stable
+          - beta
+          - dev
+          - master
+      flutter-cache:
+        description: Cache
+        type: boolean
+        default: true
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:19]: Why is this necessary?

File: .github/workflows/pr-build.yml [hunk 28]
```diff
@@ -1,45 +1,122 @@
...
+            ## ⚒️ ReVanced PR Build workflow
+            
+            ${{ env.MESSAGE }}
+            
+            You can see more details in run [${{ github.run_id }}](https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }})!
+
+            ### ⚙️ Overview
+            - App flavor: ${{ inputs.app-flavour }}
+            - Flutter branch: ${{ inputs.flutter-branch }}
+            - Flutter cache: ${{ inputs.flutter-cache }}
```
Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:112]:
```suggestion
            - Branch: ${{ inputs.flutter-branch }}
            - Cache: ${{ inputs.flutter-cache }}
```

Comment (oSumAtrIX) [inline .github/workflows/pr-build.yml:112]: Any reason the cache is relevant?

File: .github/workflows/pr-build.yml (hunk 27 above)
Comment (validcube) [reply .github/workflows/pr-build.yml:19]: I added this in case someone finds it useful because sometimes I get into a problem where I need to remove the Flutter cache to build.

File: .github/workflows/pr-build.yml (hunk 28 above)
Comment (validcube) [reply .github/workflows/pr-build.yml:112]: In my opinion, the word "Branch" & "Cache" might be too ambiguous:

* *Are we building ReVanced Manager at the `beta` branch or using the Flutter beta branch to build ReVanced*

* *Are we caching both Gradle and Flutter or just Flutter?*

Comment (oSumAtrIX) [reply .github/workflows/pr-build.yml:112]: > Are we building ReVanced Manager at the beta branch or using the Flutter beta branch to build ReVanced Manager?

This question should be treated as redundant as this branch is not aware of the branch that will eventually feature the compose version

Comment (validcube) [reply .github/workflows/pr-build.yml:112]: Cache isn't really needed to be in the build summary, but I figured that would make it cool (imo) and also test some stuff.

```suggestion
            - Flutter branch: ${{ inputs.flutter-branch }}
```

Comment (validcube) [reply .github/workflows/pr-build.yml]: Resolved @ 5068f37

**Review by Ushie — APPROVED** [review]

**Comment by KnochenAssibier** [comment]
> # 👷 Improve build workflow
> 
> > [!IMPORTANT]
> > **Last updated for 1e5833dc12777cee3b202bb942b2bf367c9d0b3a otherwise notice**
> 
> `pr-build.yml` will only be triggered when you manually dispatch it, it will comment on the PR about the build status.
> 
> the CI will require you to input your PR number, and you can customise the app flavour and flutter variant
> 
> By the way, since we are only manually dispatching it, we can finally use Gradle cache, there's less chance to hit the cache eviction compared to our old build CI.
> 
> ## Overview
> The way this work is once dispatched, the CI will build the application with the choice that you've provided, when the process of building the application is done, it will notify the owner of the Pull Request to tell them if there's a problem with their contributions or not.
> 
> ### Artefact
> If the application is built without any issues. The CI will upload the APK and SHA hash as artefact under `revanced-manager` App name will always come before the bracket, the bracket is a group that indicates everything inside was build-related, and the outside was tool-related.
> 
> **revanced-manager-(e8f044d-17-release)-stable**
> 
> image
> 
> ### Notify
> image
> 
> The CI will comment on the PR notifying them of: the time when the workflow was dispatched, configuration and build status, 
> > ## ⚒️ ReVanced PR Build workflow
> >
> > ✅ The CI has sucessfully built ReVanced Manager at 85608fc.
> >
> > You can check the details on run ID [6667156871](example.com)!
> >
> > ### ⚙️ Overview
> > - App flavor: release
> > - Flutter branch: stable
> > - Flutter cache: true
> > - Start time: Fri, 27 Oct 2023 12:25:53 +0000
> 
> the CI will also try to recreate the comment on every dispatch.
> 
> > [!WARNING]
> > Don't run the CI more than two instances of the workflow on the same PR or else it might create additional comment instead of recreating it.
> 
> > [!NOTE]
> > It's possible to use custom account such as bot account to comment by replacing the GitHub Actions token with your own.
> 
> pr-build.yml : `notify-pull-request` Job
> ```yml
> ...
>       - name: "Comment to Pull Request #${{ inputs.pr-number }}"
>         uses: thollander/actions-comment-pull-request@v2
>         with:
>           GITHUB_TOKEN: ${{ secret.bot-token }}
> ...
> ```
> 
> 
> ## Inputs
> 
> image
> 
> ### Supported Flutter branch // `choice`
> * Stable
> * Beta
> * Dev
> * Master (why)
> 
> ### Supported Flutter Cache // `boolean`
> > [!NOTE]
> > Disable when in doubt.
> * true
> * false
> 
> ### Supported app flavour // `choice`
> * Release
> * Debug
> * [Profile ⧉](https://docs.flutter.dev/perf/ui-performance#run-in-profile-mode:~:text=Flutter%E2%80%99s%20profile%20mode%20compiles%20and%20launches%20your%20application%20almost%20identically%20to%20release%20mode%2C%20but%20with%20just%20enough%20additional%20functionality%20to%20allow%20debugging%20performance%20problems.%20For%20example%2C%20profile%20mode%20provides%20tracing%20information%20to%20the%20profiling%20tools. "Flutter Profile mode")
> 
> ### Pull Request number // `string`
> > [!WARNING]
> > The field actually only accepts `integer` format.
> 
> Pull Request number without the hashtag (i.e 1337, not #1337)
> 

gh repo clone ReVanced/.github
